### DISCUSSION TITLE
How to save doc to disk with nlp.pipe(as.tuple = True)?

### DISCUSSION THREAD

**Discussion by fatihbozdag** [explosion/spaCy#11474] [Help: Coding & Implementations] [answered]
Greetings,

I am looking for a way to save `doc `file with context included. 

for task `doc = nlp(text)`, dumbing with pickle works fine. However, I want to include context data with the tagged file. 

So the question is, is there a way to save both context and `doc` in a container? for `doc, context in nlp.pipe(text, as_tuple = True)`? 
It really takes so much to process a large corpus with metadata(context) again and again for various tasks.

**Comment by fatihbozdag** [comment]
Also I've faced with a strange behaviour (it might be also due to my lack of knowledge)

for instance,

```
import spacy
import pandas as pd

nlp = spacy.load('en_core_web_lg')

meta  = pd.read_csv('.../metadata.csv')
text = pd.read_csv('.../text.csv')
meta_x = meta.to_dict('records')
text_only = text['text_field'].values.tolist()
corpus = list(zip(text_only, meta_x))
verbs = []
meta1= []
meta2 = []

annotated_text = nlp.pipe(corpus, as_tuples = True:)

for doc, context in annotated_text:
             for token in doc:
                   if token.pos_ == 'VERB':
                       verbs.append(token.lemma_)
                       meta1.append(context['META1'])
```

This works like a charm! However, if I run the loop (with the very same code) for the second time, I got no results at all. Indeed, it returns blank. 

For instance; 

```
for doc, context in annotated_text:
             for token in doc:
                   if token.pos_ == 'NOUN':
                       verbs.append(token.lemma_)
                       meta1.append(context['META1'])
```

or 

```
for doc, context in annotated_text:
             for token in doc:
                   if token.pos_ == 'VERB':
                       verbs.append(token.lemma_)
                       meta1.append(context['META2'])
```
Provides no results. So I have to rerun `annotated_text = nlp.pipe(corpus, as_tuples = True:)` so that I can run the loop again. Why is the case?

**Reply by polm** [reply]
`nlp.pipe` is what is known as a "generator" - it is like a list that is created only when you ask for the next element. This makes it take less memory, and it doesn't run computation until you need it, but a side effect is that it can only be iterated once. See [the Python docs](https://docs.python.org/3/glossary.html#term-generator) for more info.

**Comment by polm** [comment, marked as answer]
The right way to do this depends on your access pattern. 

A simple thing you can do is, if the data can be serialized easily, you can put it in a custom extension attribute (`_` attribute) and use a DocBin to serialize things. However DocBins aren't designed for random access.

If you need random access, you'll want to serialize Docs individually, so you could just create a tuple like `(doc, stuff)` and serialize it, with some kind of separate index structure. However that will be very inefficient due to the way Vocabs work, so if you need to do that you should copy just the parts of the Doc you need into another data structure.

If you give more details about the type of data you have in "context" and what your access needs are we may be able to give better advice.

**Reply by fatihbozdag** [reply]
context section includes both numeric and categorical data, however, to simplify things, if I can integrate `'docid_field' `to the spacy `doc,` it will be sufficient. Indeed, the corpus is a csv file where each column is a part of metadata, and the final column includes text for the corresponding document. There are almost 10k individual documents. 

```
data.frame':	9529 obs. of  21 variables:
 $ docid_field                       : chr  "BGSU1001" "BGSU1002" "BGSU1003" "BGSU1004" ...
 $ Native_language                   : chr  "Bulgarian" "Bulgarian" "Bulgarian" "Bulgarian" ...
 $ Country                           : chr  "Bulgaria" "Bulgaria" "Bulgaria" "Bulgaria" ...
 $ Gender                            : chr  "Female" "Female" "Female" "Female" ...
 $ Age                               : int  20 20 20 20 21 21 21 21 21 21 ...
 $ X1st_other_foreign_language       : chr  "Spanish" "Spanish" "German" "German" ...
 $ X2nd_other_foreign_language       : chr  "Russian" "Russian" NA NA ...
 $ X3rd_other_foreign_language       : chr  NA NA NA NA ...
 $ X1st_language_at_home             : chr  "Bulgarian" "Bulgarian" "Bulgarian" "Bulgarian" ...
 $ X2nd_language_at_home             : chr  NA NA NA NA ...
 $ X3rd_language_at_home             : chr  NA NA NA NA ...
 $ Years_of_English_at_school        : chr  "6 or more" "6 or more" "6 or more" "6 or more" ...
 $ Years_of_English_at_university    : chr  "2" "2" "2" "2" ...
 $ Months_in_English.speaking_country: chr  "0" "0" "0" "0" ...
 $ Title                             : chr  "Some people say that in our modern world, dominated by science and technology and industrialisation, there is n"| __truncated__ "Most University degrees are theoretical and do not prepare us for the real life. Do you agree or disagree?" "Some people say that in our modern world, dominated by science and technology and industrialisation, there is n"| __truncated__ "Most University degrees are theoretical and do not prepare us for the real life. Do you agree or disagree?" ...
 $ Length_in_words                   : int  505 504 779 525 581 580 582 533 374 326 ...
 $ Type                              : chr  "Argumentative" "Argumentative" "Argumentative" "Argumentative" ...
 $ Conditions                        : chr  "No timing" "No timing" "No timing" "No timing" ...
 $ Reference_tools                   : chr  "Yes" "Yes" "Yes" "Yes" ...
 $ Examination                       : chr  "No" "No" "No" "No" ...
 $ text_field                        : chr  " <ICLE-BG-SUN-0001.1> \nIt is time, that our society is dominated by industrialization. The prosperity of a cou"| __truncated__ " <ICLE-BG-SUN-0002.1> \nNowadays there is a growing interest in higher education and almost every young person "| __truncated__ " <ICLE-BG-SUN-0003.1> \nOnce upon a time there lived a beautiful young lady and a handsome young gentleman. The"| __truncated__ " <ICLE-BG-SUN-0004.1> \nOur educational system in one of the largest in the world. It offers variety, diversity"| __truncated__ ...
```

**Reply by polm** [reply]
In that case it sounds like an [extension attribute on the Doc](https://spacy.io/usage/processing-pipelines/#custom-components-attributes) will do what you need then.

**Reply by fatihbozdag** [reply]
I have failed to apply custom extension on  `doc` object. I couldn't figure out how to import `'docid_field'` column from pandas to `getter `function. 

`TypeError: Argument 'other' has incorrect type (expected spacy.tokens.token.Token, got str)`

So should the attribute be included in the text to be parsed? What I am trying to do is to set Doc._. from `df['docid_field']` while parsing the text included in `df['text_field']`. From the error above, it seems like the custom extension should be a `token` object from `doc`, is it true?

**Reply by kinghuang** [reply]
We also deal with a large amount of metadata in our pipeline. It looks like the documentation hasn't been updated to reflect this, but the [`Language.__call__`](https://spacy.io/api/language#call) method was modified to accept a [`Doc`](https://spacy.io/api/doc) object in addition to `str`. You can leverage this to create a Doc object for your text, add metadata to extensions on the doc, then run the doc through the pipeline.

**Register extensions first:**

```python
from spacy.tokens import Doc

# Register extensions to hold metadata.
Doc.set_extension("docid", default=None)
Doc.set_extension("language", default=None)
```

**Set extension values on doc before running through the pipeline:**

```python
text = "Some text to process"
docid_field = "BGSU1001"
native_language = "Bulgarian"

# Prepare the doc with its metadata.
doc = nlp.make_doc(text)
doc._.docid = docid_field
doc._.language = native_language

# Run the doc through the language pipeline.
doc = nlp(doc)
```

**Reply by fatihbozdag** [reply]
Awesome. How about iterating pandas list or dataframe? Csv file consists of 9529 obs, i.e. documents

**Reply by pmbaumgartner** [reply]
@fatihbozdag Usually the best way is through `df.itertuples()`, for example:

```python
for row in df.itertuples():
   doc = nlp.make_doc(row.text)
   # custom attribute logic here
```
